### PR TITLE
Add head_mask/decoder_head_mask for BART

### PR DESCRIPTION
**Pull request by stancld**
This PR implement `head_mask` and `decoder_head_mask` for PyTorch BART-based models. The full list, please, see below:
- **BART**
- **MBart**
- **Blenderbot**
- **BlenderbotSmall**
- **Marian**
- **Pegasus**

This PR is a follow up on the closed PR #9404.

**Motivation**:
According to HuggingFace's websites "There is a growing field of study concerned with investigating the inner working of large-scale transformers like BERT (that some call “BERTology”)." This PR enables to mask attention heads in encoder and decoder models exactly like for BERT. This PR thus creates an opportunity to study the importance of attention heads in encoder-decoder BERT-like model.

**Description**
- New arguments `head_mask` and`decoder_head_mask` are passed to all the BART-based models `...Model`, `...ForConditionalGeneration` and `...ForQuestionAnswering` after four arguments `input_ids, attention_mask, decoder_input_ids, decoder_attention_mask` so that a testing and whole pipeline remains smooth.

- This PR also contains updated `test_headmasking`, which currently works fine with one problem - BART-based models do not satisfy a condition:
```
self.assertNotEqual(attentions[1][..., 0, :, :].flatten().sum().item(), 0.0).

```
Fixing this problem is currently underway. 

**Reviewer:** @patrickvonplaten

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
Thanks for opening a new PR. Let me know if you need a review (It's also ok if I go into the PR and fix some things if your stuck :-) )

**Comment by stancld** [comment]
@patrickvonplaten I hope this PR is again ready for review. The only thing remaining to resolve is that issue in `test_headmasking` described above. Currently, I've been trying to fix this one, but I'll be grateful for sure if you can have a look at that too :)

**Comment by stancld** [comment]
Hey @patrickvonplaten. I would like to inform you I fixed `test_headmasking` for BART-based. The problem was that code inside
```
self.assertNotEqual(attentions[1][..., 0, :, :].flatten().sum().item(), 0.0)
```
pointed to the last layer of encoder/decoder (encoder-decoder models have only 2 layers in each module while BERT has 5 layers during testing). At the end of the day, this condition was invalid for BART-based models considering the `head_mask` to be
```
head_mask = torch.ones(
    self.model_tester.num_hidden_layers,
    self.model_tester.num_attention_heads,
    device=torch_device,
)
head_mask[0, 0] = 0
head_mask[-1, :-1] = 0
```

I hope this PR is then ready for review.

File: tests/test_modeling_common.py [hunk 1]
```diff
@@ -474,24 +482,31 @@ def test_headmasking(self):
...
-
             self.assertIsNotNone(multihead_outputs)
             self.assertEqual(len(multihead_outputs), self.model_tester.num_hidden_layers)
-            self.assertAlmostEqual(attentions[0][..., 0, :, :].flatten().sum().item(), 0.0)
-            self.assertNotEqual(attentions[0][..., -1, :, :].flatten().sum().item(), 0.0)
-            self.assertNotEqual(attentions[1][..., 0, :, :].flatten().sum().item(), 0.0)
-            self.assertAlmostEqual(attentions[-1][..., -2, :, :].flatten().sum().item(), 0.0)
-            self.assertNotEqual(attentions[-1][..., -1, :, :].flatten().sum().item(), 0.0)
+
+            def check_attentions_validity(attentions):
```
Comment (patrickvonplaten) [inline tests/test_modeling_common.py:488]: that's awesome!

File: tests/test_modeling_bart.py [hunk 2]
```diff
@@ -393,7 +402,7 @@ class BartModelTest(ModelTesterMixin, GenerationTesterMixin, unittest.TestCase):
     all_generative_model_classes = (BartForConditionalGeneration,) if is_torch_available() else ()
     is_encoder_decoder = True
     test_pruning = False
-    test_head_masking = False
+    test_head_masking = True
```
Comment (patrickvonplaten) [inline tests/test_modeling_bart.py:405]: nice!

File: tests/test_modeling_common.py [hunk 3]
```diff
@@ -465,6 +468,11 @@ def test_headmasking(self):
             head_mask.requires_grad_(requires_grad=True)
             inputs = self._prepare_for_class(inputs_dict, model_class).copy()
             inputs["head_mask"] = head_mask
+            if model.config.is_encoder_decoder:
+                signature = inspect.signature(model.forward)
+                arg_names = [*signature.parameters.keys()]
+                if "decoder_head_mask" in arg_names:  # necessary diferentiation because of T5 model
```
Comment (patrickvonplaten) [inline tests/test_modeling_common.py:474]: good for me!

**Review by patrickvonplaten — APPROVED** [review]
This is awesome! Amazing work @stancld - very clean and nice comments. If you want it would be awesome if you could open an issue regarding T5 having only a `"head_mask"`, but no `"decoder_head_mask"` and ping me on it so that we can fix this in a follow-up PR :-)

Otherwise, LGTM!

File: src/transformers/models/bart/modeling_bart.py [hunk 4]
```diff
@@ -357,6 +378,10 @@ def forward(
             encoder_hidden_states (:obj:`torch.FloatTensor`): cross attention input to the layer of shape `(seq_len, batch, embed_dim)`
             encoder_attention_mask (:obj:`torch.FloatTensor`): encoder attention mask of size
                 `(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
+            layer_head_mask (:obj:`torch.FloatTensor`): mask for attention heads in a given layer of size
+                `(config.encoder_attention_heads,)`.
+            encoder_layer_head_mask (:obj:`torch.FloatTensor`): mask for encoder attention heads in a given layer of
+            size `(config.encoder_attention_heads,)`.
```
Comment (sgugger) [inline src/transformers/models/bart/modeling_bart.py:384]:
```suggestion
                size `(config.encoder_attention_heads,)`.
```

File: src/transformers/models/blenderbot/modeling_blenderbot.py [hunk 5]
```diff
@@ -359,6 +380,10 @@ def forward(
             encoder_hidden_states (:obj:`torch.FloatTensor`): cross attention input to the layer of shape `(seq_len, batch, embed_dim)`
             encoder_attention_mask (:obj:`torch.FloatTensor`): encoder attention mask of size
                 `(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
+            layer_head_mask (:obj:`torch.FloatTensor`): mask for attention heads in a given layer of size
+                `(config.encoder_attention_heads,)`.
+            encoder_layer_head_mask (:obj:`torch.FloatTensor`): mask for encoder attention heads in a given layer of
+            size `(config.encoder_attention_heads,)`.
```
Comment (sgugger) [inline src/transformers/models/blenderbot/modeling_blenderbot.py:386]:
```suggestion
                size `(config.encoder_attention_heads,)`.
```

File: src/transformers/models/blenderbot_small/modeling_blenderbot_small.py [hunk 6]
```diff
@@ -357,6 +378,10 @@ def forward(
             encoder_hidden_states (:obj:`torch.FloatTensor`): cross attention input to the layer of shape `(seq_len, batch, embed_dim)`
             encoder_attention_mask (:obj:`torch.FloatTensor`): encoder attention mask of size
                 `(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
+            layer_head_mask (:obj:`torch.FloatTensor`): mask for attention heads in a given layer of size
+                `(config.encoder_attention_heads,)`.
+            encoder_layer_head_mask (:obj:`torch.FloatTensor`): mask for encoder attention heads in a given layer of
```
Comment (sgugger) [inline src/transformers/models/blenderbot_small/modeling_blenderbot_small.py:383]:
```suggestion
                encoder_layer_head_mask (:obj:`torch.FloatTensor`): mask for encoder attention heads in a given layer of
```

File: src/transformers/models/marian/modeling_marian.py [hunk 7]
```diff
@@ -373,6 +394,10 @@ def forward(
             encoder_hidden_states (:obj:`torch.FloatTensor`): cross attention input to the layer of shape `(seq_len, batch, embed_dim)`
             encoder_attention_mask (:obj:`torch.FloatTensor`): encoder attention mask of size
                 `(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
+            layer_head_mask (:obj:`torch.FloatTensor`): mask for attention heads in a given layer of size
+                `(config.encoder_attention_heads,)`.
+            encoder_layer_head_mask (:obj:`torch.FloatTensor`): mask for encoder attention heads in a given layer of
+            size `(config.encoder_attention_heads,)`.
```
Comment (sgugger) [inline src/transformers/models/marian/modeling_marian.py:400]:
```suggestion
                size `(config.encoder_attention_heads,)`.
```

File: src/transformers/models/mbart/modeling_mbart.py [hunk 8]
```diff
@@ -364,6 +385,10 @@ def forward(
             encoder_hidden_states (:obj:`torch.FloatTensor`): cross attention input to the layer of shape `(seq_len, batch, embed_dim)`
             encoder_attention_mask (:obj:`torch.FloatTensor`): encoder attention mask of size
                 `(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
+            layer_head_mask (:obj:`torch.FloatTensor`): mask for attention heads in a given layer of size
+                `(config.encoder_attention_heads,)`.
+            encoder_layer_head_mask (:obj:`torch.FloatTensor`): mask for encoder attention heads in a given layer of
+            size `(config.encoder_attention_heads,)`.
```
Comment (sgugger) [inline src/transformers/models/mbart/modeling_mbart.py:391]:
```suggestion
                size `(config.encoder_attention_heads,)`.
```

File: src/transformers/models/pegasus/modeling_pegasus.py [hunk 9]
```diff
@@ -374,6 +395,10 @@ def forward(
             encoder_hidden_states (:obj:`torch.FloatTensor`): cross attention input to the layer of shape `(seq_len, batch, embed_dim)`
             encoder_attention_mask (:obj:`torch.FloatTensor`): encoder attention mask of size
                 `(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
+            layer_head_mask (:obj:`torch.FloatTensor`): mask for attention heads in a given layer of size
+                `(config.encoder_attention_heads,)`.
+            encoder_layer_head_mask (:obj:`torch.FloatTensor`): mask for encoder attention heads in a given layer of
+            size `(config.encoder_attention_heads,)`.
```
Comment (sgugger) [inline src/transformers/models/pegasus/modeling_pegasus.py:401]:
```suggestion
                size `(config.encoder_attention_heads,)`.
```

**Review by sgugger — APPROVED** [review]
Looks great! Thanks for adding this!

File: tests/test_modeling_common.py [hunk 10]
```diff
@@ -474,24 +482,31 @@ def test_headmasking(self):
...
+                    self.assertLess(
+                        torch.sum(torch.isnan(t)), t.numel() / 4
+                    )  # Check we don't have more than 25% nans (arbitrary)
+                attentions = [
+                    t.masked_fill(torch.isnan(t), 0.0) for t in attentions
+                ]  # remove them (the test is less complete)
+
+                self.assertAlmostEqual(attentions[0][..., 0, :, :].flatten().sum().item(), 0.0)
+                self.assertNotEqual(attentions[0][..., -1, :, :].flatten().sum().item(), 0.0)
+                if len(attentions) > 2:  # encoder-decodere models have only 2 layers in each modules
```
Comment (LysandreJik) [inline tests/test_modeling_common.py:500]:
```suggestion
                if len(attentions) > 2:  # encoder-decoder models have only 2 layers in each module
```

**Review by LysandreJik — APPROVED** [review]
This is great, very clean implementation! Thanks for implementing the tests, too.

LGTM!
